### ISSUE TITLE
Dependencies: deprecate `setup.py` and move to `pyproject.toml`

### ISSUE THREAD

**Issue by rdimaio** [rucio/rucio#6971]
`pyproject.toml` is the current specification standard for build system requirements for Python projects, defined in PEP-518: https://peps.python.org/pep-0518/#file-format

`setup.py` is not deprecated, but it's recommended to move to `pyproject.toml` if possible:
- https://packaging.python.org/en/latest/discussions/setup-py-deprecated/
- https://stackoverflow.com/questions/75588250/what-are-the-downsides-to-relying-purely-on-pyproject-toml
- https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#is-pyproject-toml-mandatory

To do:
- [ ] Implement this first: https://github.com/rucio/containers/issues/344 (we want to more accurately replicate the package installation process in tests)
- [ ] Follow https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#modernize-setup-py-project to see what can be moved to `pyproject.toml`

**Comment by mgajek-cern** [comment]
Has a tradeoff analysis been done between using `setuptools` in `pyproject.toml` versus adopting tools like `Poetry`, `Hatch` or others? It would be helpful to understand the rationale, especially considering their benefits in dependency locking, environment management and modern workflows.

Official Python Packaging Authority (PyPA) guidance: https://packaging.python.org/en/latest/guides/tool-recommendations/

**Comment by rdimaio** [comment]
> Has a tradeoff analysis been done between using `setuptools` in `pyproject.toml` versus adopting tools like `Poetry`, `Hatch` or others? It would be helpful to understand the rationale, especially considering their benefits in dependency locking, environment management and modern workflows.
> 
> Official Python Packaging Authority (PyPA) guidance: https://packaging.python.org/en/latest/guides/tool-recommendations/

I can see the idea in migrating to a more modern tool like Poetry, but I think it would be hard to do that with Rucio at this stage. Rucio packaging is still done in a fairly manual way until we complete the namespace package restructure, which would allow us to actually build the different packages separately without needing to rename the setup script everytime. You're welcome to investigate the migration, but I don't think it's worth considering this for now.

**Comment by Geogouz** [comment]
According to the linked PR #7675, this issue shall be handled by that. However, there are TODOs here in the description that are pointing to other PRs (which are delayed). I assume those TODOs are just related to this issue and not required for addressing it. Right?

**Comment by rdimaio** [comment]
> According to the linked PR [#7675](https://github.com/rucio/rucio/pull/7675), this issue shall be handled by that. However, there are TODOs here in the description that are pointing to other PRs (which are delayed). I assume those TODOs are just related to this issue and not required for addressing it. Right?

Yes, I think that at least the first TODO can be tracked under its separate existing issue
